### PR TITLE
Add Giphy integration

### PR DESCRIPTION
**Pull request by Kegsay**

### CODE REVIEW

File: src/github.com/matrix-org/go-neb/matrix/matrix.go [hunk 1]
```diff
@@ -116,6 +126,52 @@ func (cli *Client) SendText(roomID, text string) (string, error) {
...
+// UploadLink uploads an HTTP URL and then returns an MXC URI.
+func (cli *Client) UploadLink(link string) (string, error) {
+	res, err := http.Get(link)
+	if res != nil {
+		defer res.Body.Close()
+	}
+	if err != nil {
+		return "", err
+	}
+	log.Print("GOT GIF BODY ", link)
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/go-neb/matrix/matrix.go]: THIS IS EXTREMELY EXCITING LOGGING!

File: src/github.com/matrix-org/go-neb/matrix/matrix.go [hunk 2]
```diff
@@ -116,6 +126,52 @@ func (cli *Client) SendText(roomID, text string) (string, error) {
...
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("Content-Type", contentType)
+	cl, err := strconv.Atoi(contentLength)
+	if err != nil {
+		return "", err
+	}
+	req.ContentLength = int64(cl)
+	log.Print("Doing upload request to ", req.URL)
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/go-neb/matrix/matrix.go]: Maybe want to be using log.WithFields here?

File: src/github.com/matrix-org/go-neb/matrix/matrix.go [hunk 3]
```diff
@@ -116,6 +126,52 @@ func (cli *Client) SendText(roomID, text string) (string, error) {
...
+func (cli *Client) UploadLink(link string) (string, error) {
+	res, err := http.Get(link)
+	if res != nil {
+		defer res.Body.Close()
+	}
+	if err != nil {
+		return "", err
+	}
+	log.Print("GOT GIF BODY ", link)
+	return cli.UploadToContentRepo(res.Body, res.Header.Get("Content-Type"), res.Header.Get("Content-Length"))
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/go-neb/matrix/matrix.go]: https://golang.org/pkg/net/http/#Response has a ContentLength field fwiw

File: src/github.com/matrix-org/go-neb/matrix/matrix.go [hunk 4]
```diff
@@ -116,6 +126,52 @@ func (cli *Client) SendText(roomID, text string) (string, error) {
...
+}
+
+// UploadToContentRepo uploads the given bytes to the content repository and returns an MXC URI.
+func (cli *Client) UploadToContentRepo(content io.Reader, contentType, contentLength string) (string, error) {
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/go-neb/matrix/matrix.go]: I'd prolly take contentLength as an int64 here.

File: src/github.com/matrix-org/go-neb/services/giphy/giphy.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	}
+}
+func (s *giphyService) cmdGiphy(client *matrix.Client, roomID, userID string, args []string) (interface{}, error) {
+	// only 1 arg which is the text to search for.
+	query := strings.Join(args, " ")
+	gifResult, err := s.searchGiphy(query)
+	if err != nil {
+		return nil, err
+	}
+	log.Print("GOT ", gifResult)
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/go-neb/services/giphy/giphy.go]: MORE EXCITING LOGGING!

File: src/github.com/matrix-org/go-neb/matrix/matrix.go (hunk 1 above)
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/matrix/matrix.go]: oops

Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/matrix/matrix.go]: done

File: src/github.com/matrix-org/go-neb/matrix/matrix.go (hunk 2 above)
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/matrix/matrix.go]: done

File: src/github.com/matrix-org/go-neb/matrix/matrix.go (hunk 3 above)
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/matrix/matrix.go]: done

File: src/github.com/matrix-org/go-neb/matrix/matrix.go (hunk 4 above)
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/matrix/matrix.go]: done

File: src/github.com/matrix-org/go-neb/services/giphy/giphy.go (hunk 5 above)
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/services/giphy/giphy.go]: done

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**